### PR TITLE
fix(NumberInput): leading zeros removed from inputs

### LINKED ISSUE #7117: Bug - NumberInput - Leading zeros
Leading zeros should be stripped, for example if the user types `0`, then adds another positive integer like `3`, the input should strip the leading zero and display `3` only.

Expected:
1. Type `405`
2. Delete first digit `4`
3. Input should update to show the value `5`

Actual:
1. Type `405`
2. Delete first digit `4`
3. Input shows `05`

Looks like if the final value assigned to the input in the jsx is a string, it works fine, but right now we're passing a number to it.

Current:
```
<input value={value} />
```

Possible solution:
```
<input value={Number(value).toString()} />
```

Related issue:
https://github.com/facebook/react/issues/9402

### PR DESCRIPTION
**Pull request by andyyvo**
**What**: Closes #7117 . Leading zero is removed from `NumberInput` component by implementing an `onBlur` prop for all inputs. If the user types 0, then adds another positive integer like 3, the input now strips the leading zero and display 3 only. This works for negative values and for decimal places.

**Additional issues**: The only issue was the **With Unit** component examining monetary amounts currently takes in a string and was thus, not implemented with the leading zero. Furthermore, the `toFixed(2)` function should probably be called on `onChange` as well as to prevent user inputs of anything passed two decimals.

### CODE REVIEW

File: packages/react-core/src/components/NumberInput/examples/NumberInput.md [hunk 1]
```diff
@@ -37,6 +37,10 @@ class BasicNumberInput extends React.Component {
         value: this.state.value + 1
       });
     };
+
+    this.onBlur = (e) => {
+      e.target.value = Number(e.target.value).toString();
```
Comment (jenny-s51) [inline packages/react-core/src/components/NumberInput/examples/NumberInput.md:42]: It looks like we want to directly update the `value` state variable - this aligns with the implementation of all the other event handlers. 

Can we use state wherever `this.onBlur` is declared in `NumberInput.md`? 

```suggestion
      const newValue = Number(e.target.value).toString();
       this.setState({
         value: newValue
       });
```

**Review by jenny-s51 — CHANGES_REQUESTED** [review]
Looking good @andyyvo - nice work; inputs are behaving as intended now. Left a comment about `onBlur`, will be just a small update! 

Otherwise LGTM 🙂

File: packages/react-core/src/components/NumberInput/examples/NumberInput.md (hunk 1 above)
Comment (andyyvo) [reply packages/react-core/src/components/NumberInput/examples/NumberInput.md:42]: Good catch! I'll add that to all of the `onBlur` instances :)

Comment (andyyvo) [reply packages/react-core/src/components/NumberInput/examples/NumberInput.md:42]: We might not need to update the state in `onBlur` since we do that in `onChange` and console logging `value` in both `onChange` and `onBlur` appear to have the leading zero cut. I'm going to keep looking into this!

Comment (jenny-s51) [reply packages/react-core/src/components/NumberInput/examples/NumberInput.md:42]: Ah you're right! Did some investigating here too... since we need to keep the `toString()` for the leading zeros fix, I'm OK approving as-is -- matches how @jschuler wrote it in #7117 👍

Comment (andyyvo) [reply packages/react-core/src/components/NumberInput/examples/NumberInput.md:42]: Sounds great! I just made sure that `onBlur` was also added to the second example with the monetary amount

**Review by nicolethoen — COMMENTED** [review]
Okay - if this should be the default behavior of the NumberInput, should we bring the `Number(e.target.value).toString();` logic into the component so the consumer doesn't have to implement this fix themself?

Something sort of like how the `onChange` prop is passed to the `TextInput` component and then called from an internal function (which could also include your extra logic) which is ultimately what's passed to the input element.

**Review by mcarrano — APPROVED** [review]
Looks good from a UX perspective. Thanks @andyyvo !

**Comment by andyyvo** [comment]
> Okay - if this should be the default behavior of the NumberInput, should we bring the `Number(e.target.value).toString();` logic into the component so the consumer doesn't have to implement this fix themself?
> 
> Something sort of like how the `onChange` prop is passed to the `TextInput` component and then called from an internal function (which could also include your extra logic) which is ultimately what's passed to the input element.

+1 on this -- how would I go about changing the component itself to implement this logic?

**Comment by jenny-s51** [comment]
@andyyvo you can try to put the `.toString()` conversion directly on `value` here: https://github.com/patternfly/patternfly-react/blob/main/packages/react-core/src/components/NumberInput/NumberInput.tsx#L117

**Comment by andyyvo** [comment]
Shout to @nicolethoen for this one! Just added a fix to **NumberInput.tsx** where a `handleBlur` function now renders the logic after any `onChange` or `onBlur` event so that the consumer no longer would have to write out this logic themselves. @jenny-s51 if you could take a look -- that would be great!

**Review by nicolethoen — APPROVED** [review]

**Review by tlabaj — CHANGES_REQUESTED** [review]
Can you please add unit or integration test to verify the behaviour for the different scenarios (negative number, decimals, etc).

File: packages/react-core/src/components/NumberInput/NumberInput.tsx [hunk 2]
```diff
@@ -87,6 +87,16 @@ export const NumberInput: React.FunctionComponent<NumberInputProps> = ({
   const keyDownHandler =
     inputProps && inputProps.onKeyDown ? inputProps.onKeyDown : defaultKeyDownHandler({ inputName, onMinus, onPlus });
 
+  const handleBlur = (event: React.FocusEvent<HTMLInputElement>) => {
+    event.target.value = Number(event.target.value).toString();
```
Comment (tlabaj) [inline packages/react-core/src/components/NumberInput/NumberInput.tsx:91]: I would create a local event variable and set it to the event then modify that.
```suggestion
    let ev = event.target.value;
    ev.target.value = Number(event.target.value).toString();
```

File: packages/react-core/src/components/NumberInput/NumberInput.tsx [hunk 3]
```diff
@@ -87,6 +87,16 @@ export const NumberInput: React.FunctionComponent<NumberInputProps> = ({
...
+    if (onChange) {
+      onChange(event);
```
Comment (tlabaj) [inline packages/react-core/src/components/NumberInput/NumberInput.tsx:93]:
```suggestion
      onChange(ev);
```

File: packages/react-core/src/components/NumberInput/NumberInput.tsx [hunk 4]
```diff
@@ -87,6 +87,16 @@ export const NumberInput: React.FunctionComponent<NumberInputProps> = ({
...
+    }
+    if (onBlur) {
+      onBlur(event);
```
Comment (tlabaj) [inline packages/react-core/src/components/NumberInput/NumberInput.tsx:96]:
```suggestion
      onBlur(ev);
```

**Review by tlabaj — CHANGES_REQUESTED** [review]

File: packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx [hunk 5]
```diff
@@ -60,4 +61,90 @@ describe('numberInput', () => {
...
+
+  test('renders unit with negative decimal number', () => {
+    const { asFragment } = render(<NumberInput value={-12.4} unit="%" unitPosition="after" />);
+    expect(asFragment()).toMatchSnapshot();
+  });
+
+  test('renders unit & position with negative decimal number', () => {
+    const { asFragment } = render(<NumberInput value={-12.4} unit="$" unitPosition="before" />);
+    expect(asFragment()).toMatchSnapshot();
+  });
```
Comment (wise-king-sullyman) [inline packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx:108]: @tlabaj are these tests for all these different possible combinations of input values and units something you were looking for?

I feel like the tests that I've highlighted here could be trimmed down to just a test for `unit` with `"before"` and a test for `unit` with `"after"` since the unit and its position aren't dependent on what the input value is.

Comment (tlabaj) [reply packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx:108]: @wise-king-sullyman I think that is fine to trim that down.  I was not specific enough with my comment.
I was actually looking for test to verify the trimming of the leading zeros works with different values ( decimal, negative numbers) as stated in description.

Comment (andyyvo) [reply packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx:108]: @wise-king-sullyman @tlabaj I'll remove the tests on units and position then and it should be all set for review

Comment (wise-king-sullyman) [reply packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx]: @andyyvo I think that @tlabaj is looking for tests which verify that the new behavior with leading zeros being removed works as expected, which I'm not currently seeing in the test suite. Please correct me if I'm wrong about my read of your last comment @tlabaj.

To do so I'm guessing you'll need to use `userEvent` to type a leading zero into the input then click outside of the input, then make an assertion about the value of the input, rather than just being able to pass it into `NumberInput` as a value.

Comment (tlabaj) [reply packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx]: @wise-king-sullyman that is correct.  This could also be done with a cypress integration test.

**Comment by andyyvo** [comment]
Struggling to write a test for leading zeros. Here are all the tests I've tried that have failed:

```
test('removes leading zero', () => {
    render(<NumberInput value={0320} />); // this is an octal literal which is its own error
    expect(screen.getByDisplayValue('320')).toBeInTheDocument();

    const input = screen.getByRole('spinbutton'); // tried textbox and it doesn't work
    userEvent.type(input, '0'); // if initial value is 320, it remains 320

    expect(screen.getByRole('spinbutton')).toHaveValue('0320');
});
```

```
test('removes leading zero', () => {
    const val = (123).toString().padStart(5, '0'); // the issue here is that value takes in a number and not string
    render(<NumberInput value={parseInt(val)} />); // parseInt defeats the purpose of the test
    expect(screen.getByDisplayValue('00123')).toBeInTheDocument();
});
```

File: packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx [hunk 6]
```diff
@@ -60,4 +61,82 @@ describe('numberInput', () => {
...
+  test('removes leading zero', () => {
+    render(<NumberInput value={10} onChange={() => {}} />);
+  
+    const input = screen.getByRole('spinbutton');
+    userEvent.type(input, '{arrowleft}{arrowleft}0');
+    expect(input).toHaveDisplayValue('010');
+  
+    userEvent.click(document.body)
+  
+    expect(input).toHaveDisplayValue('10');
```
Comment (wise-king-sullyman) [inline packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx:140]: Can you add tests to verify that the leading zeros are removed on decimal and negative numbers as well?

**Review by wise-king-sullyman — CHANGES_REQUESTED** [review]
Looking good! I think there are a few other behaviors to test still though.

File: packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx [hunk 7]
```diff
@@ -60,4 +61,128 @@ describe('numberInput', () => {
...
+
+  test('removes trailing zero', () => {
+    render(<NumberInput value={120.0} />);
+    expect(screen.getByDisplayValue('120')).toBeInTheDocument();
+  });
+
+  test('removes many trailing zeroes', () => {
+    render(<NumberInput value={120.0100} />);
+    expect(screen.getByDisplayValue('120.01')).toBeInTheDocument();
+  });
```
Comment (wise-king-sullyman) [inline packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx:129]: I don't think that these trailing zero tests are needed because I believe number inputs have this behavior by default.

File: packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx [hunk 8]
```diff
@@ -60,4 +61,128 @@ describe('numberInput', () => {
...
+
+  test('removes a trailing zero from a negative number', () => {
+    render(<NumberInput value={-40.0} />);
+    expect(screen.getByDisplayValue('-40')).toBeInTheDocument();
+  });
+
+  test('removes trailing zeroes from a negative number', () => {
+    render(<NumberInput value={-40.0100} />);
+    expect(screen.getByDisplayValue('-40.01')).toBeInTheDocument();
+  });
```
Comment (wise-king-sullyman) [inline packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx:151]: See my previous comment regarding trailing zero testing.

File: packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx [hunk 9]
```diff
@@ -60,4 +61,128 @@ describe('numberInput', () => {
...
+    render(<NumberInput value={-55.0} onChange={() => {}} />);
+  
+    const input = screen.getByRole('spinbutton');
+    userEvent.type(input, '{arrowleft}{arrowleft}0');
+    expect(input).toHaveDisplayValue('-055');
+  
+    userEvent.click(document.body)
+  
+    expect(input).toHaveDisplayValue('-55');
+  });
```
Comment (wise-king-sullyman) [inline packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx:187]: I don't believe these tests are needed either since the leading zero removal behavior is already covered in the preceding tests.

File: packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx [hunk 10]
```diff
@@ -60,4 +61,128 @@ describe('numberInput', () => {
...
 });
```
Comment (wise-king-sullyman) [inline packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx:188]: Can you add a test for leading zero removal for decimal numbers?

**Review by wise-king-sullyman — CHANGES_REQUESTED** [review]

File: packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx (hunk 10 above)
Comment (andyyvo) [reply packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx:188]: Do we not need to do the same tests with negative numbers?

Comment (wise-king-sullyman) [reply packages/react-core/src/components/NumberInput/__tests__/NumberInput.test.tsx:188]: Yes, keep the leading zero removal test for negative numbers that you currently have, or do you mean a test for negative decimal numbers?

**Review by wise-king-sullyman — APPROVED** [review]
Looks great! Thanks for all your work on this and especially these tests 🚀

**Review by tlabaj — APPROVED** [review]
LGTM! Nice work!

**Review by jenny-s51 — APPROVED** [review]
🚀

**Review by nicolethoen — APPROVED** [review]
